man/tilestache-compose.1
.TH TILESTACHE-COMPOSE 1 "Nov 10, 2010"
.SH NAME
tilestache-compose \- set a map coverage area
.SH SYNOPSIS
.B tilestache-compose
.RI [ options ] " file" ...
.SH DESCRIPTION
This manual page documents briefly the \fBtilestache-compose\fR command.
.PP
There are three ways to set a map coverage area.
.PP
1) Center, zoom, and dimensions: create a map of the specified size,
centered on a given geographical point at a given zoom level:
.PP
   tilestache-compose.py \-c config.json \-l layer-name \-d 800 800 \-n 37.8 \-122.3 \-z 11 out.jpg
.PP
2) Extent and dimensions: create a map of the specified size that
adequately covers the given geographical extent:
.PP
   tilestache-compose.py \-c config.json \-l layer-name \-d 800 800 \-e 36.9 \-123.5 38.9 \-121.2 out.png
.PP
3) Extent and zoom: create a map at the given zoom level that covers
the precise geographical extent, at whatever pixel size is necessary:
.PP
   tilestache-compose.py \-c config.json \-l layer-name \-e 36.9 \-123.5 38.9 \-121.2 \-z 9 out.jpg
.SH OPTIONS
.TP
\fB-c\fR, \fB\-\-config\fR \fIfile\fR
Path to configuration file.
.TP
\fB-l\fR, \fB\-\-layer\fR \fIlayer\fR
Layer name from configuration.
.TP
\fB-n\fR, \fB\-\-center\fR \fIlat lon\fR
Geographic center of map.
.TP
\fB-e\fR, \fB\-\-extent\fR \fIlat lon lat lon\fR
Geographic extent of map.
.TP
\fB-z\fR, \fB\-\-zoom\fR \fIlevel\fR
Zoom level.
.TP
\fB-d\fR, \fB\-\-dimensions\fR \fIwidth height\fR
Pixel width, height of output image.
.TP
\fB-v\fR, \fB\-\-verbose\fR
Make a bunch of noise.
.TP
\fB-i\fR, \fB\-\-include-paths\fR \fIpaths\fR
Add the following colon-separated list of paths to Python's include path (aka sys.path)
.TP
\fB-x\fR, \fB\-\-ignore-cached\fR
Re-render every tile, whether it is in the cache already or not.
.TP
\fB-h\fR, \fB\-\-help\fR
Show summary of options.
.SH AUTHOR
\fBTileStache\fR was written by Michal Migurski <mike@stamen.com>.


CHANGELOG
2016-12-21: 1.51.4
- Support for GDAL 2.x types OFTInteger64 and OFTInteger64 on Vector Tiles

2016-10-30: 1.51.3
- Fixes for debian packaging

2016-10-27: 1.51.2
- Adding support for .pbf format for Mapbox Vector Tiles
- Configuration can be passed as a dictionary in addition to path to JSON file.

2016-09-01: 1.51.1
- Bugfix on Mapnik image object

2016-09-01: 1.51
- Replacing PIL by Pillow
- Added example with raster

2015-01-30: 1.50.1
- Fixed missing ttf file

2015-01-21: 1.50.0
- Added 'scale factor' option to Mapnik image provider
- Added S3 policy specification to config
- Added dictionary config to VecTiles config
- Added pixel effects to all layer configurations
- Added remote configuration to MapnikGrid to match Mapnik
- Fixed UtfGridComposite to work with latest TileStache
- Fixed Pillow dependency
- Fixed "cache lifespan" in Redis cache

2014-06-07: 1.49.11
- Fixed 'allow-origin header' getting overwritten (JesseCrocker)
- Added 'source projection' parameter to UrlTemplate provider (JesseCrocker)
- Fixed minor AttributeError when palette256 argument is None (PerryGeo)
- Added Timeout feature to Proxy and UrlTemplate providers (Juliomalegria)
- Adjusted curl examples (Peter Richardson)
- More specific error message when SpatialRef is missing (probabble)

2014-05-10: 1.49.10
- Fixed Travis build.
- Fixed import errors for case-insensitive filesystems.
- Added TileStache Vagrant machine configuration.
- Fixed some memcache testing problems.

2014-05-10: 1.49.9
- Moved everything to PyPI and fixed VERSION kerfuffle.

2013-07-02: 1.49.8
- Dropped Proxy provider srs=900913 check.
- Updated most JSON mime-types from text/json to application/json.
- Fixed requestHandler() query_string argument to match existing API docs.
- Switched order of simplify and transform in VecTiles queries to handle GEOS bug.
- Made Rummy CORS-compliant.

2013-06-23: 1.49.7
- Added VecTiles:MultiProvider to merge TopoJSON and GeoJSON output layers.

2013-06-21: 1.49.6
- Switched to ST_SimplifyPreserveTopology in VecTiles clipping to calm GEOS.

2013-06-20: 1.49.5
- Tweaked VecTiles Provider documentation.

2013-06-20: 1.49.4
- Added TopoJSON support to VecTiles Provider.

2013-06-15: 1.49.3
- Fixed runaway query bug in VecTiles Provider with looped bounds in query_columns().

2013-06-14: 1.49.2
- Fixed SQL bug in VecTiles Provider for limiting column queries.

2013-06-14: 1.49.1
- Updated magic column names in VecTiles Provider to __geometry__ and __id__ for feature ID.

2013-06-05: 1.49.0
- Added option to use S3 reduced redundancy storage for cheaper caching.
- Made Postgres connections per-request instead of long term in VecTiles Provider.

2013-05-28: 1.48.2
- Fixed a serious documentation incompatibility with TileStache.requestHandler() introduced in 1.47.0.

2013-05-27: 1.48.1
- Fixed Werkzeug dependency in setup.py.

2013-05-26: 1.48.0
- Fixed a serious documentation incompatibility with TileStache.getTile() introduced in 1.47.0.

2013-04-24: 1.47.4
- Fixed from-mbtiles/output-directory use case in tilestache-seed, which I had broken.

2013-04-22: 1.47.3
- Fixed clobbered content from TheTileLeftANote exception.

2013-04-22: 1.47.2
- Fixed getTile bugs introduced in 1.47.0.

2013-04-20: 1.47.1
- Fixed typo in setup.py.

2013-04-20: 1.47.0
- Exposed mutable HTTP status codes and headers to TileStache.getTile() and caches via TheTileLeftANote exception.
- Removed old layer.redirects implementation via TheTileIsInAnotherCastle exception.
- Improved PIL dependency in setup requirements to include Pillow, with patch from Fabian Büchler.
- Fixed coordinate generator usage in tilestache-seed with patch from Seth Fitzsimmons.
- Added Redis cache provider with patch from Matthew Perry.

2013-04-12: 1.46.10
- Prefixed most Postgis functions in VecTile provider with "ST_" for forward compatibility.

2013-04-12: 1.46.9
- Added ogr.OFTDate and ogr.OFTTime to list of acceptable field types in Vector provider.

2013-04-12: 1.46.8
- Added ogr.OFTDateTime to list of acceptable field types in Vector provider, with patch from probabble.

2013-03-27: 1.46.7
- Fixed a stupid error in VecTiles.server discarding important Postgres connection details.

2013-03-26: 1.46.6
- Improved error tolerance in VecTiles client and server.

2013-03-25: 1.46.5
- Added new simplify_until parameter to VecTiles.Provider.
- Added some invalid geometry checks to VecTiles.server.
- Added !bbox! token to VecTiles query parser.

2013-03-19: 1.46.4
- Added new zoom_data parameter to TileStache.Goodies.VecTiles Datasource.

2013-03-19: 1.46.3
- Added new simplify parameter to TileStache.Goodies.VecTiles Provider.

2013-03-14: 1.46.2
- Added new parameters to TileStache.Goodies.VecTiles Datasource and Provider.

2013-03-14: 1.46.1
- Fixed missing TileStache.Goodies.VecTiles by adding it to setup.py.

2013-03-12: 1.46.0
- Added TileStache.Goodies.VecTiles to support efficient vector data tiles for use in Mapnik.
- Fixed alpha-channel issue in Sandwich provider with patch from Lars Ahlzen.

2013-02-26: 1.45.0
- Made S3 credentials optional since Boto allows them in via environment vars or config file.
- Added path parameter to S3 cache, to use a subdirectory rather than bucket root.
- Eased bitmap size checks in Sandwich provider for retina use.
- Added "quadtile" directory layout for disk cache provider.
- Added JSON support to MBTiles provider from Aaron Cope.
- Added Travis CI tests for Vector layer.

2013-02-22: 1.44.0
- Merged Ragi Burhum's Travis continuous integration tests with ~1/3rd test coverage.
- Modified Proxy provider to return untouched original image if size and type match.
- Updated Mapnik usage to reflect recently released versions.
- Fixed Mapnik provider deadlock bug with patch from Lee Shepherd.
- Added "key prefix" argument to Memcache cache provider.

2013-01-03: 1.43.0
- Added new tilestache-list.py script to feed tilestache-seed.py --tile-list.

scripts/tilestache-compose.py
#!/usr/bin/env python
from sys import stderr, path
from tempfile import mkstemp
from thread import allocate_lock
from os import close, write, unlink
from optparse import OptionParser
from os.path import abspath

import ModestMaps

mmaps_version = tuple(map(int, getattr(ModestMaps, '__version__', '0.0.0').split('.')))

if mmaps_version < (1, 3, 0):
    raise ImportError('tilestache-compose.py requires ModestMaps 1.3.0 or newer.')

#
# More imports can be found below, after the --include-path option is known.
#

class Provider (ModestMaps.Providers.IMapProvider):
    """ Wrapper for TileStache Layer objects that makes them behave like ModestMaps Provider objects.

        Requires ModestMaps 1.3.0 or better to support "file://" URLs.
    """
    def __init__(self, layer, verbose=False, ignore_cached=None):
        self.projection = layer.projection
        self.layer = layer
        self.files = []

        self.verbose = bool(verbose)
        self.ignore_cached = bool(ignore_cached)
        self.lock = allocate_lock()

        #
        # It's possible that Mapnik is not thread-safe, best to be cautious.
        #
        self.threadsafe = self.layer.provider is not TileStache.Providers.Mapnik

    def tileWidth(self):
        return 256

    def tileHeight(self):
        return 256

    def getTileUrls(self, coord):
        """ Return tile URLs that start with file://, by first retrieving them.
        """
        if self.threadsafe or self.lock.acquire():
            mime_type, tile_data = TileStache.getTile(self.layer, coord, 'png', self.ignore_cached)

            handle, filename = mkstemp(prefix='tilestache-compose-', suffix='.png')
            write(handle, tile_data)
            close(handle)

            self.files.append(filename)

            if not self.threadsafe:
                # must be locked, right?
                self.lock.release()

            if self.verbose:
                size = len(tile_data) / 1024.
                printlocked(self.lock, self.layer.name() + '/%(zoom)d/%(column)d/%(row)d.png' % coord.__dict__, '(%dKB)' % size)

            return ('file://' + abspath(filename), )

    def __del__(self):
        """ Delete any tile that was saved in self.getTileUrls().
        """
        for filename in self.files:
            unlink(filename)

class BadComposure(Exception):
    pass

def printlocked(lock, *stuff):
    """
    """
    if lock.acquire():
        print ' '.join([str(thing) for thing in stuff])
        lock.release()

parser = OptionParser(usage="""tilestache-compose.py [options] file

There are three ways to set a map coverage area.

1) Center, zoom, and dimensions: create a map of the specified size,
   centered on a given geographical point at a given zoom level:

   tilestache-compose.py -c config.json -l layer-name -d 800 800 -n 37.8 -122.3 -z 11 out.jpg

2) Extent and dimensions: create a map of the specified size that
   adequately covers the given geographical extent:

   tilestache-compose.py -c config.json -l layer-name -d 800 800 -e 36.9 -123.5 38.9 -121.2 out.png

3) Extent and zoom: create a map at the given zoom level that covers
   the precise geographical extent, at whatever pixel size is necessary:

   tilestache-compose.py -c config.json -l layer-name -e 36.9 -123.5 38.9 -121.2 -z 9 out.jpg""")

defaults = dict(center=(37.8044, -122.2712), zoom=14, dimensions=(900, 600), verbose=True)

parser.set_defaults(**defaults)

parser.add_option('-c', '--config', dest='config',
                  help='Path to configuration file.')

parser.add_option('-l', '--layer', dest='layer',
                  help='Layer name from configuration.')

parser.add_option('-n', '--center', dest='center', nargs=2,
                  help='Geographic center of map. Default %.4f, %.4f.' % defaults['center'], type='float',
                  action='store')

parser.add_option('-e', '--extent', dest='extent', nargs=4,
                  help='Geographic extent of map. Two lat, lon pairs', type='float',
                  action='store')

parser.add_option('-z', '--zoom', dest='zoom',
                  help='Zoom level. Default %(zoom)d.' % defaults, type='int',
                  action='store')

parser.add_option('-d', '--dimensions', dest='dimensions', nargs=2,
                  help='Pixel width, height of output image. Default %d, %d.' % defaults['dimensions'], type='int',
                  action='store')

parser.add_option('-v', '--verbose', dest='verbose',
                  help='Make a bunch of noise.',
                  action='store_true')

parser.add_option('-i', '--include-path', dest='include_paths',
                  help="Add the following colon-separated list of paths to Python's include path (aka sys.path)")

parser.add_option('-x', '--ignore-cached', action='store_true', dest='ignore_cached',
                  help='Re-render every tile, whether it is in the cache already or not.')

if __name__ == '__main__':

    (options, args) = parser.parse_args()

    if options.include_paths:
        for p in options.include_paths.split(':'):
            path.insert(0, p)

    import TileStache

    try:
        if options.config is None:
            raise TileStache.Core.KnownUnknown('Missing required configuration (--config) parameter.')

        if options.layer is None:
            raise TileStache.Core.KnownUnknown('Missing required layer (--layer) parameter.')

        config = TileStache.parseConfig(options.config)

        if options.layer not in config.layers:
            raise TileStache.Core.KnownUnknown('"%s" is not a layer I know about. Here are some that I do know about: %s.' % (options.layer, ', '.join(sorted(config.layers.keys()))))

        provider = Provider(config.layers[options.layer], options.verbose, options.ignore_cached)

        try:
            outfile = args[0]
        except IndexError:
            raise BadComposure('Error: Missing output file.')

        if options.center and options.extent:
            raise BadComposure("Error: bad map coverage, center and extent can't both be set.")

        elif options.extent and options.dimensions and options.zoom:
            raise BadComposure("Error: bad map coverage, dimensions and zoom can't be set together with extent.")

        elif options.center and options.zoom and options.dimensions:
            lat, lon = options.center[0], options.center[1]
            width, height = options.dimensions[0], options.dimensions[1]

            dimensions = ModestMaps.Core.Point(width, height)
            center = ModestMaps.Geo.Location(lat, lon)
            zoom = options.zoom

            map = ModestMaps.mapByCenterZoom(provider, center, zoom, dimensions)

        elif options.extent and options.dimensions:
            latA, lonA = options.extent[0], options.extent[1]
            latB, lonB = options.extent[2], options.extent[3]
            width, height = options.dimensions[0], options.dimensions[1]

            dimensions = ModestMaps.Core.Point(width, height)
            locationA = ModestMaps.Geo.Location(latA, lonA)
            locationB = ModestMaps.Geo.Location(latB, lonB)

            map = ModestMaps.mapByExtent(provider, locationA, locationB, dimensions)

        elif options.extent and options.zoom:
            latA, lonA = options.extent[0], options.extent[1]
            latB, lonB = options.extent[2], options.extent[3]

            locationA = ModestMaps.Geo.Location(latA, lonA)
            locationB = ModestMaps.Geo.Location(latB, lonB)
            zoom = options.zoom

            map = ModestMaps.mapByExtentZoom(provider, locationA, locationB, zoom)

        else:
            raise BadComposure("Error: not really sure what's going on.")

    except BadComposure, e:
        print >> stderr, parser.usage
        print >> stderr, ''
        print >> stderr, '%s --help for possible options.' % __file__
        print >> stderr, ''
        print >> stderr, e
        exit(1)

    if options.verbose:
        print map.coordinate, map.offset, '->', outfile, (map.dimensions.x, map.dimensions.y)

    map.draw(False).save(outfile)
